compiler/rustc_ast_lowering/src/pat.rs

```diff
@@ -69,7 +69,16 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                             ImplTraitContext::Disallowed(ImplTraitPosition::Path),
                             None,
                         );
-                        break hir::PatKind::Path(qpath);
+                        let kind = hir::PatExprKind::Path(qpath);
+                        let span = self.lower_span(pattern.span);
+                        let expr = hir::PatExpr { hir_id: pat_hir_id, span, kind };
+                        let expr = self.arena.alloc(expr);
+                        return hir::Pat {
+                            hir_id: self.next_id(),
+                            kind: hir::PatKind::Expr(expr),
+                            span,
+                            default_binding_modes: true,
+                        };
```

Comment:
Fishy, but I can't remember why this is the right soluation. If I do it differently we end up not being able to handle assoc consts anymore

Reply:
Is this because `ast::Pat` still has the `Patkind::Path` and `PatKind::Expr(ExprKind::Path` duplication, I would expect name resolve for `PatKind::Path` to store the resolutions associated with the nodeid of the pat whereas name resolution of `PatKind::Expr(ExprKind::Path` to be associated with the nodeid of the expr.

So, if you don't swap the hirids here you wind up trying to lower a hir path from a nodeid that has no resolutions associated with it.

Is it possible to remove this duplication in the ast level aswell as in the hir level? i.e. remove `ast::PatKind::Path`

Reply:
(I assume that's probably an involved change though, so can just add a comment explaining why this is the way it is and move on if  you'd like)

Reply:
> So, if you don't swap the hirids here you wind up trying to lower a hir path from a nodeid that has no resolutions associated with it.

Yea, that's it.



> I assume that's probably an involved change though,

Not really, but I'd still wanna do it as a separate PR

Reply:
reasonable :+1: